### DISCUSSION TITLE
Contract fatal problem: something went wrong during prepareProductContractions

### DISCUSSION THREAD

**Discussion by armandlcz** [FeynCalc/feyncalc#204] [Q&A]
Hello FeynCalc team and FeynCalc users!.

I'm trying to compute the following contraction:

Contract[  ( A[\[Mu],\[Nu],a,b,c,d] + B[\[Mu],\[Nu],a,b,c,d] )  *  ( FVD[l,\[Rho]] FVD[l,\[Rho]] C[c,d,a,b] + B[\[Lambda],\[Delta],c,d,a,b] FVD[l,\[Lambda]] FVD[l,\[Delta]] )  ]

where the tensors A, B, C are made out of the metric tensor in D dimensions: MTD[ \[Mu] , \[Nu] ] and all their indices are Lorentz indices (even the a, b, c, d indices are Lorentz indices). When I compute it on my personal computer everything goes well, but when I try to compute it on another computer ( where I have just installed Mathematica 12.0 and FeynCalc 9.3.1) what I get is the next messagge:

'Contract: Error! Contract has encountered a fatal problem and must abort the computation. The problem reads: Something went wrong during prepareProductContractions.'

If I erase FVD[l,\[Rho]] FVD[l,\[Rho]] in the second factor then I don't get error anymore: I mean if I compute

Contract[  ( A[\[Mu],\[Nu],a,b,c,d] + B[\[Mu],\[Nu],a,b,c,d] )  *  ( C[c,d,a,b] + B[\[Lambda],\[Delta],c,d,a,b] FVD[l,\[Lambda]] FVD[l,\[Delta]] )  ]

don't get error. I have the same error when I try to compute other similar contractions. The same error appears when trying to contract many indices and only in the computer where I have just installed Mathematica (in my personal computer I don't get this error. I have to use the first computer because it's a too heavy calculation for my personal computer). I looked for a solution on the forum but I didn't find any topic about this.

I hope you can help me. Thanks a lot.

Both computers have Mathematica 12.0 and FeynCalc 9.3.1 (stable version).

**Comment by vsht** [comment]
Hi,

please provide a minimal working example that reproduces the problem. Rather than describing what A, B and C are, wouldn't it be easier
to write them down explicitly? ;)

Otherwise, it is hardly possible to understand what goes wrong here.

Cheers,
Vladyslav

**Reply by armandlcz** [reply]
Hi dear Vladyslav. 

First of all, I thank you for your reply. I didn't write A, B and C explicitly because they are huge expressions, but I tracked the error to simpler expressions as the following ones:

Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] 
-> 'Contract: Error! Contract has encountered a fatal problem and must abort the computation. The problem reads: Something went wrong during prepareProductContractions.'

This is a minimal working example that reproduces the problem. If I take off any of the factors inside Contract or even any term inside any of these factors then I don't get error anymore. (I know you might say there is no need to place FVD[p, \[Mu]] inside Contract because index \[Mu] is a free index, not a contracted one, and then I should place it out of Contract. I know, but in the original expression it's not possible to factor out free indices). What's more, If I add a factor FVD[p, \[Mu]] in order to contract with itself then I don't get error:

Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] 
-> No error

This is a little list of some expressions that throw error and some other ones that don't:
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> error
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) MTD[ \[Mu], \[Nu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> error
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> NO error
Contract[(MTD[r, a] MTD[s, b] ) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> NO error
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b])] -> NO error
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> NO error
Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a])] -> NO error

and the next ones are maybe the most striking ones:

Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] FVD[l, \[Lambda]] FVD[l, \[Lambda]] + MTD[r, b] MTD[s, a] FVD[l, \[Lambda]] FVD[l, \[Lambda]])] -> error

Contract[(MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[p, \[Mu]] (MTD[r, a] MTD[s, b] + MTD[r, b] MTD[s, a]) FVD[l, \[Lambda]] FVD[l, \[Lambda]] ] -> NO error

I think they're striking because it's the same mathematical expression just written in a different way (the second one is just a factorization of the first one).

The same happens if I use 4-dimensional functions instead of D-dimensional ones: MTD[]->MT[] and FVD[]->FV[]. As I said before, I get these errors in one computer but not in the other one I have. Both have same version of Mathematica and Feyncalc. I already tried to uninstall and reinstall Mathematica and Feyncalc and also applied update from Win10 to Win11 and all Windows updates available, but they didn't solve my problem.

Well, I would thank you a lot if you can give me any light to solve this issue. If you need more information I would be pleased to give it to you. Finally, I want to thank you for Feyncalc. It's a wonderful tool!. 

Best regards,
armandlcz.

**Reply by vsht** [reply]
Hi,

many thanks for the detailed examples. I've just pushed a fix to the dev version, could you please check that
it solves all the relevant issues?

After that I would add it to the stable branch as well.

Cheers,
Vladyslav

**Reply by armandlcz** [reply]
Hi dear Vladyslav.

Yes!, the issue is solved in the dev version!. Thank you a lot!.

By the way, thanks and congrats for FeynCalc, it's a wonderful tool!.

**Reply by vsht** [reply]
Many thanks for the confirmation!

Cheers,
Vladyslav
